sklearn/linear_model/cd_fast.pyx

```diff
@@ -265,33 +294,30 @@ def enet_coordinate_descent(np.ndarray[DOUBLE, ndim=1] w,
                     const = 1.0
                     gap = R_norm2
 
-                l1_norm = dasum(n_features, <DOUBLE*>w.data, 1)
+                l1_norm = asum(n_features, w_data, 1)
 
                 # np.dot(R.T, y)
-                gap += (alpha * l1_norm - const * ddot(
-                            n_samples,
-                            <DOUBLE*>R.data, 1,
-                            <DOUBLE*>y.data, n_tasks)
+                gap += (alpha * l1_norm
+                        - const * dot(n_samples, R_data, 1, y_data, n_tasks)
                         + 0.5 * beta * (1 + const ** 2) * (w_norm2))
 
                 if gap < tol:
                     # return if we reached desired tolerance
                     break
-
     return w, gap, tol, n_iter + 1
 
 
 @cython.boundscheck(False)
 @cython.wraparound(False)
 @cython.cdivision(True)
-def sparse_enet_coordinate_descent(double[:] w,
-                            double alpha, double beta,
-                            np.ndarray[double, ndim=1, mode='c'] X_data,
+def sparse_enet_coordinate_descent(floating [:] w,
+                            floating alpha, floating beta,
+                            np.ndarray[floating, ndim=1, mode='c'] X_data,
                             np.ndarray[int, ndim=1, mode='c'] X_indices,
                             np.ndarray[int, ndim=1, mode='c'] X_indptr,
-                            np.ndarray[double, ndim=1] y,
-                            double[:] X_mean, int max_iter,
-                            double tol, object rng, bint random=0,
+                            np.ndarray[floating, ndim=1] y,
+                            floating[:] X_mean, int max_iter,
+                            floating tol, object rng, bint random=0,
```

Comment:
Yes, but then we may need to make `w_max` and other related variables become non-floating according to [this line](https://github.com/scikit-learn/scikit-learn/pull/6913/files#diff-5d8b82d506a543bd684a1d403d5941e0R267). (**Note**: `d_w_tol` = `tol`)

And since `w_max` is used to compare with `w` (which must be `floating`) at [this line](https://github.com/scikit-learn/scikit-learn/pull/6913/files#diff-5d8b82d506a543bd684a1d403d5941e0R263), things will getting too complicated in my opinion.